src/renderer/base/renderer.cpp

```diff
@@ -627,32 +621,19 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
             screenPoint.X += gsl::narrow<SHORT>(cols);
             cols = 0;
 
-            // Ensure that our cluster vector is clear.
-            clusters.clear();
-
             // This inner loop will accumulate clusters until the color changes.
             // When the color changes, it will save the new color off and break.
             do
             {
-                if (color != it->TextAttr())
-                {
-                    color = it->TextAttr();
-                    break;
-                }
-
-                // Walk through the text data and turn it into rendering clusters.
-                clusters.emplace_back(it->Chars(), it->Columns());
-
                 // Advance the cluster and column counts.
-                const auto columnCount = clusters.back().GetColumns();
-                it += columnCount > 0 ? columnCount : 1; // prevent infinite loop for no visible columns
+                const auto columnCount = (*clusterIter).GetColumns();

```

Comment:
Calculating the `cols` is cumbersome. , I don't know how to do this in a more elegant way.

Reply:
Oh. This might actually justify making the `PaintBufferLine`'s first arg be a & instead.

Save a copy of it on the way in, let the engine increment it, then add a method that lets you difference the two of them.

Diff the resultant one you get back versus the one you initially created and it should tell you how many columns it walked. Then you don't have to cycle through the loop multiple times.

That's what I did for the cell/text iterators with `.GetCellDistance` and whatnot.

Reply:
I'm not sure. I don't think it's enough to justify this. Making it `&` sounds very error-prone and implicit.

Reply:
Make two of them? Take a const input and then have a second & as the output one?

It's not uncommon to use an _Inout_ parameter, but if you think its safer to have an explicit _In_ and _Out_ (given we can't use the return code as it's always `HRESULT` for these thing), it seems OK.